azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -0,0 +1,165 @@
+            self._engine.add_policy(rule_file)
+            self._engine.add_data(policy_file)
+            self._policy_engine_enabled = True
+        except Exception as ex:
+            msg = "Failed to initialize Regorus policy engine: '{0}'".format(ex)
+            self.log_policy(msg=msg)
+            raise PolicyError(msg)
+
+    @classmethod
+    def log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_event=True):
+        """
+        Log information to console and telemetry.
+        """
+        if is_success:
+            logger.info("[Policy] " + msg)
+        else:
+            logger.error("[Policy] " + msg)
+        if send_event:
+            add_event(op=op, message=msg, is_success=is_success)
+
+    @staticmethod
+    def is_policy_enforcement_enabled():
+        """Check whether user has opted into policy enforcement feature"""
+        # TODO: Remove conf flag post private preview and add other checks here.
+        return conf.get_extension_policy_enabled()
+
+    @staticmethod
+    def is_policy_enforcement_supported():
+        """Check that both platform architecture and distro/version are supported."""
+        osutil = get_osutil()
+        arch = osutil.get_vm_arch()
+        if arch not in POLICY_SUPPORTED_ARCHITECTURE:
+            return False
+        __distro__ = get_distro()
+        DISTRO_NAME = __distro__[0]
+        DISTRO_VERSION = __distro__[1]
+        try:
+            distro_version = DistroVersion(DISTRO_VERSION)
+        except ValueError:
+            raise ValueError
+
+        # Check if the distro is in the support matrix and if the version is supported
+        if DISTRO_NAME in POLICY_SUPPORTED_DISTROS_MIN_VERSIONS:
+            min_version = POLICY_SUPPORTED_DISTROS_MIN_VERSIONS[DISTRO_NAME]
+            return distro_version >= min_version
+        else:
+            return False
+
+    @property
+    def policy_engine_enabled(self):
+        """This property tracks whether the feature is enabled and Regorus engine has been successfully initialized"""
+        return self._policy_engine_enabled
+
+    def evaluate_query(self, input_json, query):
+        """
+        Expected format for input_json:
+        {
+            "extensions": {
+                "<extension_name_1>": {
+                    "signingInfo": {
+                        "extensionSigned": <true, false>
+                    }
+                }, ...
+        }
+
+        Expected format for query: "data.agent_extension_policy.extensions_to_download"
+        """
+        # This method should never be called if policy is not enabled, this would be a developer error.
+        if not self.policy_engine_enabled:
+            raise PolicyError("Policy enforcement is disabled, cannot evaluate query.")
+
+        try:
+            self._engine.set_input(input_json)
+            full_result = self._engine.eval_query(query)
+            value = full_result['result'][0]['expressions'][0]['value']
```

Comment:
How does the output of the engine when the query indicates an extension violates a policy? what is the exit code in that case? the error exist codes documented in regorus.py seem to indicate more issues in the inputs to the engine than extensions not complying with policy.

Reply:
The engine will execute successfully (no error). "value" will look something like: 

"Microsoft.Azure.ActiveDirectory.AADSSHLoginForLinux": {
       "downloadAllowed": true 
}, 
"Microsoft.Azure.Extensions.CustomScript": { 
       "downloadAllowed": false
}

ExtensionPolicyEngine.is_extension_download_allowed( ) will be responsible for parsing this output and returning true/false based on the value of "downloadAllowed". I've added examples to the spec doc that will hopefully make this easier to understand.

Reply:
thanks for the sample

the code above should check that 'result', 'expressions' and 'value' are in the output and issue an appropriate error message

Reply:
added this validation